File: 0151-reverse-words-in-a-string/0151-reverse-words-in-a-string.py
class Solution:
    def reverseWords(self, s: str) -> str:
        n = len(s)
        j = n - 1

        while j >= 0 and s[j] == " ":
            j-=1

        start = j

        j = 0

        while j <= n - 1 and s[j] == " ":
            j += 1
        
        end = j

        if start < end:
            return ""

        ans = []
        word = []

        for i in range(start, end - 1, -1):
            if s[i] != " ":
                word.append(s[i])
            else:
                if s[i] == s[i + 1]:
                    continue

                while word:
                    ans.append(word.pop())

                ans.append(" ")

        while word:
            ans.append(word.pop())

        return "".join(ans)

File: 0151-reverse-words-in-a-string/0151-reverse-words-in-a-string.cpp
class Solution {
public:
    string reverseWords(string s) {
        int n = s.size();
        int j = n - 1;

        while (j >= 0 && s[j] == ' ')
            j--;

        int start = j;

        j = 0;

        while (j < n && s[j] == ' ')
            j++;

        int end = j;

        if (start < end)
            return "";

        string ans = "";
        vector<char> word;

        for (int i = start; i >= end; i--) {
            if (s[i] != ' ') {
                word.push_back(s[i]);
            } else {
                if (s[i] == s[i + 1])
                    continue;

                while (!word.empty()) {
                    ans += word.back();
                    word.pop_back();
                }

                ans += ' ';
            }
        }

        while (!word.empty()) {
            ans += word.back();
            word.pop_back();
        }

        return ans;
    }
};